File: 1209-Remove All Adjacent Duplicates in String II.py
class Solution(object):
    def removeDuplicates(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        stack = []
        for i in range(len(s)):
            if not stack or s[i] != stack[-1][0]:
                stack.append([s[i], 1])
            else:
                if stack[-1][1] + 1 == k:
                    stack.pop()
                else:
                    stack[-1][1] += 1
        
        res = ""
        for x in stack:
            res += x[1]*x[0]
        return res

File: JavaScript/1209-Remove All Adjacent Duplicates in String II.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var removeDuplicates = function(s, k) {
    let stack = [];
    for (let i=0; i<s.length; i++){
        if (!stack.length || s[i] != stack[stack.length-1][0]){
            stack.push([s[i], 1])
        }else{
            if ( stack[stack.length-1][1] + 1 === k){
                stack.pop();
            }else{
                stack[stack.length-1][1] += 1;
            }
        }
    }

    res = ""
    for (let x of stack){
        for (let i=0; i<x[1]; i++){
            res += x[0];
        }
    }
    return res;
};